torchvision/prototype/features/_feature.py

```diff
@@ -234,3 +234,7 @@ def gaussian_blur(self, kernel_size: List[int], sigma: Optional[List[float]] = N
 
 InputType = Union[torch.Tensor, PIL.Image.Image, _Feature]
 InputTypeJIT = torch.Tensor
+
+
+class GenericFeature(_Feature):
```

Comment:
Now named `GenericDatapoint`, but the spirit is the same.

Reply:
IIUC, this extra class is needed solely because pure Tensors are currently treated as Images, for BC reason.

Perhaps we can rely on a different convention which is that if there are pure Tensors in a sample, only the *first* pure Tensor will be treated as an Image if there's not already an Image; and all the other Tensors will be passed-through.

I feel like this convention would work for every single datasets we currently provide (or we can manually enforce it in the dataset -> transforms wrapper.

For user-defined datasets it should work for the majority of cases. For the cases where it doesn't work, well... Those cases didn't work for our current transforms anyway - so there's no BC consideration to have here.

Whether we would still need a `GenericDatapoint` or allow the base `DataPoint` class to be pass-through is up for discussion. My initial gut feeling would be to not expose that right away and wait for users feedback, but happy to hear others' thoughts on this

Reply:
To clarify, you propose to replace

https://github.com/pytorch/vision/blob/135a0f9ea9841b6324b4fe8974e2543cbb95709a/torchvision/prototype/transforms/_transform.py#L40-L42

with something along the lines:

```python
flat_outputs = []
image_or_video_found = False
for inpt in flat_inputs:
    if check_input(inpt, self._transformed_types):
        if check_input(inpt, datapoints.Image, is_simple_tensor, PIL.Image.Image):
            if image_found:
                output = inpt
            else:
                image_found = True
                output = self._transform(inpt, params)
        else:
            output = self._transform(inpt, params)
    else:
        output = inpt
        
    flat_outputs.append(output)
```

(maybe we can untangle that a little, but the functionality should be clear)

---

Assuming the above is what you meant, I can't say I'm a fan. I agree that it will work for our datasets and most of the current custom ones as well. However, part of the v2 API is the ability to perform the transformation on arbitrary input structures. And this somewhat violates that, since now the order plays a role again. So basically we'll from then on say

> Transforms v2 handles _almost_ arbitrary input structures

Again, I don't think there will be a lot of incompatibility, I just dislike that we need to walk back even slightly on this IMO very nice feature. But then I'm totally biased in that as well.

Reply:
TBH I don't really see how the "arbitrary input structure" idea is violated.

We just make an assumption on the order, just like we currently make the assumption that Tensor == Image (which, if there's more than one tensor, is the wrong assumption to make in most cases).

> Transforms v2 handles almost arbitrary input structures

I would still say that the proposal handles all input structures. What input structures aren't supported?

Reply:
> We just make an assumption on the order, just like we currently make the assumption that Tensor == Image (which, if there's more than one tensor, is the wrong assumption to make in most cases).

I agree, but when the user actually has such a case like `{"image_1": image_tensor_1, "image_2": image_tensor_2}` that won't work. Basically we are limiting the API to a single plain tensor image.

So unless I'm missing something, we need to decide whether we want to 

1. allow multiple tensor images, but users have to wrap every other tensor that is not an image in a `GenericDatapoint`, or we 
2. make the assumption that users can only pass a single plain tensor image and have to manually wrap into `datapoints.Image` if they have multiple ones.

Reply:
> but when the user actually has such a case like {"image_1": image_tensor_1, "image_2": image_tensor_2} that won't work. Basically we are limiting the API to a single plain tensor image.

Yes but I don't think this is more limiting than treating all tensors as image? For every user who wants to pass multiple tensor images, there's another user who wants to pass a single image tensor and pass-through the rest without having to bother with an extra class.

I really don't think one option is more limiting than the other. I do think however that we should avoid introducing new classes as much as possible (it bloats the users' mental model). And we should remember than any of those use-cases aren't supported by V1 right now anyway. Users who want to transforms more than one image should probably be encouraged to wrap their images as Images rather than pass them as tensors.

Reply:
We're taking about a use-case that:

- wasn't supported before; so users have to write new code, and they can probably wrap their stuff into Image as they go
- doesn't even lead to anything meaningful within torchvision right now: there's no model that cares about 2 images, apart from the OF ones which aren't supported by V2 rn

So I'm OK with not addressing this potential risk which is IMO extremely niche / low. The risk of exposing an additional class seems higher and would impact more users

Reply:
> Forgetting to wrap something into a GenericDatapoint will most likely fail loudly due to unmet assumptions that we have for tensor images and thus at least mitigates this problem.

BTW this is purely speculative but if we assume that there's a Resize in 99% of the transformation pipelines, we're almost guaranteed to get a loud error as well if only one of the input images is resized while the rest are passed-through.

Reply:
> So I'm OK with not addressing this potential risk which is IMO extremely niche / low. The risk of exposing an additional class seems higher and would impact more users

Fair enough. Last try from my side:

As mentioned in the second part of https://github.com/pytorch/vision/pull/6663#discussion_r982567437, instead of having an extra class that enables this, we could also use the `Datapoint` base class an thus not need any extra class. Given that it is currently not exposed, it would still increase our API surface. But then if we want to go for #7028 at some point, we need to expose it eventually anyway.

Reply:
> we could also use the Datapoint base class an thus not need any extra class

Trying to clarify that to make sure I understand: if we pass a DataPoint right now, it's already completely passed-through, right? I.e. we would just need to make the DataPoint class public (and document this behaviour)?

Regardless, I'd rather go with the new heuristic option right now, and wait for users feedback and see if some people find it confusing. After all, none of the options discussed so far are really mutually exclusive.

Reply:
> Trying to clarify that to make sure I understand: if we pass a DataPoint right now, it's already completely passed-through, right? I.e. we would just need to make the DataPoint class public (and document this behaviour)?

Correct. Note that passing a plain `Datapoint` will still go through the regular dispatch mechanism, but since all methods on the base class just return `self`, every dispatcher is a no-op. This is different than for example a string that will be filtered out in 

https://github.com/pytorch/vision/blob/135a0f9ea9841b6324b4fe8974e2543cbb95709a/torchvision/prototype/transforms/_transform.py#L40-L42

and thus never hits a dispatcher.

> Regardless, I'd rather go with the new heuristic option right now, and wait for users feedback and see if some people find it confusing. After all, none of the options discussed so far are really mutually exclusive.

Ok then. Let me send another PR since that behavior is only partially related to this PR.